### PR TITLE
added random fuzziness to twap amount

### PR DESCRIPTION
**Pull request by avsek477**

### CODE REVIEW

File: lib/twap/util/generate_order.js [hunk 1]
```diff
@@ -20,13 +22,22 @@ const { DUST } = Config
...
   const {
-    sliceAmount, orderType, symbol, amount, lev, _margin, _futures
+    sliceAmount, orderType, symbol, amount, lev, _margin, _futures, amountDistortion
   } = args
+  let orderAmount
 
+  if (_isFinite(amountDistortion)) {
+    const plusOrMinus = Math.random() > 0.5 ? 1 : -1
+    const randomPercentage = ((Math.random() * amountDistortion * plusOrMinus) + 1) // random percentage from 1 to amountDistortion
+    orderAmount = sliceAmount * randomPercentage
```
Comment (robertkowalski) [inline lib/twap/util/generate_order.js:32]:
```
> 0.1 * 0.2
0.020000000000000004
```

can you use the bignumber lib we use also for iceberg orders?

File: lib/twap/util/generate_order.js [hunk 2]
```diff
@@ -1,6 +1,8 @@
 'use strict'
 
+const _isFinite = require('lodash/isFinite')
```
Comment (robertkowalski) [inline lib/twap/util/generate_order.js:3]: can you add a unit test for `generate_order.js`? should also cover the float issue

File: lib/twap/util/generate_order.js (hunk 1 above)
Comment (avsek477) [reply lib/twap/util/generate_order.js:32]: i used the prepareAmount from bfx-api-node-util library for it .. so thought not using bignumber would be okay .. but sure i will use bignumber for this

File: lib/twap/util/generate_order.js [hunk 3]
```diff
@@ -20,13 +22,22 @@ const { DUST } = Config
...
   const { args = {}, remainingAmount } = state
   const {
-    sliceAmount, orderType, symbol, amount, lev, _margin, _futures
+    sliceAmount, orderType, symbol, amount, lev, _margin, _futures, amountDistortion
   } = args
+  let orderAmount
 
+  if (_isFinite(amountDistortion)) {
+    const plusOrMinus = Math.random() > 0.5 ? 1 : -1
+    const randomPercentage = ((Math.random() * amountDistortion * plusOrMinus) + 1) // random percentage from 1 to amountDistortion
```
Comment (robertkowalski) [inline lib/twap/util/generate_order.js:31]: just to double check, lets say i choose a 2% distortion.

i have an initial value of 100 - so the submtited values are withign the range of 98 - 102, correct?

can you describe and test if everything works with the overall order amount?

e.g. i have an overall amount of 100, distortion 2%, with 5 slices...

does the algo then could overuse amounts from adding up additional distortions?

**Review by robertkowalski — CHANGES_REQUESTED** [review]
nice progress so far! needs some more testing

File: lib/twap/util/generate_order.js (hunk 3 above)
Comment (avsek477) [reply lib/twap/util/generate_order.js:31]: the submitted orders never exceeds the total amount entered by the user despite the additional distortions.. in every interval tick, if tradeBeyondEnd is enabled, the algo ensures that the next order would not push us over the total amount and doesnt submit new orders; and when tradeBeyondEnd is disabled, the remaining amount is sent for the new order when the next order pushes over the total amount 
i will write the test cases covering these possible cases

File: lib/twap/util/generate_order.js [hunk 4]
```diff
@@ -18,16 +21,25 @@ const { DUST } = Config
  * @returns {object} order - null if no amount remains
  */
 const generateOrder = (state = {}, price) => {
-  const { args = {}, remainingAmount } = state
+  const { orders = {}, args = {}, remainingAmount } = state
   const {
-    sliceAmount, orderType, symbol, amount, lev, _margin, _futures
+    sliceAmount, orderType, symbol, amount, lev, _margin, _futures, amountDistortion
   } = args
+  let orderAmount=0
```
Comment (robertkowalski) [inline lib/twap/util/generate_order.js:28]: style: `orderAmount = 0`

File: lib/twap/meta/get_ui_def.js [hunk 5]
```diff
@@ -105,6 +106,12 @@ const getUIDef = () => ({
       priceField: 'price'
     },
 
+    amountDistortion: {
+      component: 'input.percent',
+      label: 'Amount Distortion %',
+      customHelp: 'Amount to distort individual order sizes to prevent detection, in percent'
```
Comment (robertkowalski) [inline lib/twap/meta/get_ui_def.js:112]: default should be 0

**Pull request merged**